### PR TITLE
Plugin Filedelete | Catch whole file contents on NtClose (Part 1)

### PR DESCRIPTION
**Pull request by skvl**
Hello.

This a work in progress to make filedelete plugin capable to catch the whole file contents of modified or deleted files.

Current status:
- Content fetching is based on function injection on NtClose.
- File modification catching is based on trapping NtWriteFile (ZwCreateSection in future).
- [x] Only modified files are catched (deleted files will be catched in future).
- [x] Only first 16KB of file is catched (full contents catch in future).
- [x] All traps in Ring3 (Ring0 in future).
- [x] Merge into original *plugin/filedelete*.

Known issues:
~~Sometimes files not catched.~~
    Advanced research needed for `STATUS_PENDING` and `STATUS_INVALID_DEVICE_REQUEST`
~~Adwanced applications use NtCreateSection/NtMapViewOfSection for writing into files.~~
    Advanced research needed to handle `STATUS_PENDING`. Seems like file handle could be closed after section handle obtaining.

### CODE REVIEW

**Comment by skvl** [comment]
@tklengyel I think you are right. But I don't like the idea to switch context. So I'm going to finish this work.
More over I hope that after gathering a lot of examples of function injection I will be abble to make good general interface.

**Comment by skvl** [comment]
@tklengyel I have restored original filedelete plugin by renamin my implementation into filedelete2. It's disabled by default.

Could you please accept the PR?

Thanks in advance.

**Comment by tklengyel** [comment]
@skvl Thanks for this PR, it is a creative way to extend the filedelete plugin. I'm affraid in its current form I can't accept the PR. As I stated before, copying the injector code into a plugin is not good, I don't want to have multiple copies of that code laying around as it is fragile and will lead to bitrot. There are also a large number of TODO items in the code and cornercases which have not been addressed, so I'm not sure this plugin as-is is ready for general use. Also, I would prefer if this functionality was added to the existing filedelete plugin, potentially enabled with a command-line flag, instead of adding a whole new one.

**Comment by tklengyel** [comment]
Also FYI, consider reworking the code to make use of https://github.com/tklengyel/drakvuf/pull/407

**Comment by skvl** [comment]
@tklengyel could you pay some attention to the PR. I believe it is ready for now.

Though I will continue to investigate how to improve this functionality and make it even more reliable.

**Comment by tklengyel** [comment]
Is it absolutely necessary to split the filedelete plugin into two (filedelete.cpp and filedelete2.cpp)? I don't mind having multiple cpp files per plugin but if the two don't share enough code in common then they might was well be separate plugins.

**Comment by tklengyel** [comment]
@drakvuf-jenkins This is OK to test

**Comment by skvl** [comment]
@tklengyel I have tried to leave all common code inside `filedelete.cpp` and moved out some special handlers into `filedelete2`.

**Comment by skvl** [comment]
@tklengyel I don't see details of build error from *Jenkins*. But I build this branch successfully.

File: src/main.cpp [hunk 1]
```diff
@@ -177,6 +178,7 @@ int main(int argc, char** argv)
 #ifdef ENABLE_PLUGIN_FILEDELETE
                 "\t -D <file dump folder>     Folder where extracted files should be stored at\n"
                 "\t -M                        Dump new or modified files also (requires -D)\n"
+                "\t -n                        Use extraction method based on function injection\n"
```
Comment (tklengyel) [inline src/main.cpp]: Add `(requires -D)`

**Comment by tklengyel** [comment]
With that minor adjustment of the command line print helper I'm happy to go ahead and merge this.

File: src/main.cpp (hunk 1 above)
Comment (skvl) [reply src/main.cpp]: Done.

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please.

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Comment by skvl** [comment]
@tklengyel what's the issue here? I have tested this branch successfully.

**Comment by tklengyel** [comment]
Jenkins is being finicky.. Probably not related to the PR itself..

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Comment by tklengyel** [comment]
```
CID 190892 (#1 of 1): Dereference before null check (REVERSE_INULL)check_after_deref: Null-checking injector suggests that it may be null, but it has already been dereferenced on all paths leading to the check.
560    if (injector)
561        delete injector;
```

```
CID 190891 (#2 of 2): Dereference null return value (NULL_RETURNS)13. dereference: Dereferencing a pointer that might be null fp when calling fclose.
262    fclose(fp);

**Pull request merged**